src/interface/pipe.rs

```diff
@@ -89,60 +168,127 @@ impl EmulatedPipe {
     }
 
-    // Write length bytes from pointer into pipe
+    /// ### Description
+    ///
+    /// write_to_pipe writes a specified number of bytes starting at the given pointer to a circular buffer.
+    ///
+    /// ### Arguments
+    ///
+    /// write_to_pipe accepts three arguments:
+    /// * `ptr` - a pointer to the data being written.
+    /// * `length` - the amount of bytes to attempt to write
+    /// * `nonblocking` - if this attempt to write is nonblocking
+    ///
+    /// ### Returns
+    ///
+    /// Upon successful completion, the amount of bytes written is returned.
+    /// In case of a failure, an error is returned to the calling syscall.
+    ///
+    /// ### Errors
+    ///
+    /// * `EAGAIN` - Non-blocking is enabled and the write has failed to fully complete.
+    /// * `EPIPE` - An attempt to write to a pipe with all read references have been closed.
+    ///
+    /// ### Panics
+    ///
+    /// A panic occurs if the provided pointer is null
+    ///
+    /// To learn more about pipes and the write syscall
+    /// [pipe(7)](https://man7.org/linux/man-pages/man7/pipe.7.html)
+    /// [write(2)](https://man7.org/linux/man-pages/man2/write.2.html)
     pub fn write_to_pipe(&self, ptr: *const u8, length: usize, nonblocking: bool) -> i32 {
-        let mut bytes_written = 0;
-
+        // convert the raw pointer into a slice to interface with the circular buffer
         let buf = unsafe {
             assert!(!ptr.is_null());
             slice::from_raw_parts(ptr, length)
         };
 
         let mut write_end = self.write_end.lock();
-
-        let pipe_space = write_end.remaining();
-        if nonblocking && (pipe_space == 0) {
-            return syscall_error(
-                Errno::EAGAIN,
-                "write",
-                "there is no data available right now, try again later",
-            );
-        }
+        let mut bytes_written = 0;
 
         while bytes_written < length {
             if self.get_read_ref() == 0 {
+                // we send EPIPE here since all read ends are closed
                 return syscall_error(Errno::EPIPE, "write", "broken pipe");
-            } // EPIPE, all read ends are closed
+            }
 
             let remaining = write_end.remaining();
 
-            if remaining == 0 {
-                interface::lind_yield(); //yield on a full pipe
-                continue;
-            }
-            // we write if the pipe is empty, otherwise we try to limit writes to 4096 bytes (unless whats leftover of this write is < 4096)
-            if remaining != self.size
-                && (length - bytes_written) > PAGE_SIZE
-                && remaining < PAGE_SIZE
-            {
-                continue;
+            // we loop until either more than a page of bytes is free in the pipe
+            // for why we wait for a free page of bytes, refer to the pipe man page about atomicity of writes or the pipe_write kernel implementation
```

Comment:
Can you write 2-3 more sentences to explain?  Others (and you from the future) will appreciate not having to reread man pages and / or kernel source code to try to reconstruct what you were thinking.

Reply:
the manpage explanation is several paragraphs but I'll try to summarize and meet in the middle here